src/main/java/org/spdx/utility/compare/TemplateRegexGenerator.java

```diff
@@ -0,0 +1,322 @@
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.spdx.licenseTemplate.ILicenseTemplateOutputHandler;
+import org.spdx.licenseTemplate.LicenseParserException;
+import org.spdx.licenseTemplate.LicenseTemplateRule;
+import org.spdx.licenseTemplate.LicenseTemplateRuleException;
+import org.spdx.licenseTemplate.SpdxLicenseTemplateHelper;
+
+/**
+ * Generates a regular expression from a license or exception template
+ * 
+ * Note that the regular expression assumes a fully normalized text string to match
+ * 
+ * <code>findTemplateWithinText(String text)</code> will return text matching the template based on the generated start and end regexes
+ * 
+ * <code>getCompleteRegex()</code> will return a regular expression for the entire license where
+ * <code>getStartRegex(int wordLimit)</code> will return a regular expression to match the beginning of a license
+ * and <code>getEndRegex(int wordLimit)</code> will return a regular expression to match the end of a license
+ * 
+ * @author Gary O'Neall
+ *
+ */
+public class TemplateRegexGenerator implements ILicenseTemplateOutputHandler {
+	
+	static final Logger logger = LoggerFactory.getLogger(TemplateRegexGenerator.class);
+	
+	static final int WORD_LIMIT = 25; // number of words to search for at the beginning and end of the template
+	
+	static abstract class RegexElement {
+	}
+	
+	static class RegexList extends RegexElement {
+		private List<RegexElement> elements = new ArrayList<>();
+		
+		public void addElement(RegexElement element) {
+			elements.add(element);
+		}
+		
+		public List<RegexElement> getElements() {
+			return elements;
+		}
+		
+		@Override
+		public String toString() {
+			StringBuilder sb = new StringBuilder();
+			for (RegexElement element:elements) {
+				sb.append(element.toString());
+			}
+			return sb.toString();
+		}
+	}
+	
+	/**
+	 * Group with the optional modifier
+	 */
+	static class OptionalRegexGroup extends RegexList {
+		
+		@Override
+		public String toString() {
+			return "(" + super.toString() + ")?";
+		}
+	}
+	
+	static class RegexToken extends RegexElement {
+		private String token;
+		
+		public RegexToken(String token) {
+			this.token = token.trim();
+		}
+		
+		public String getToken() {
+			return token;
+		}
+		
+		@Override
+		public String toString() {
+			return Pattern.quote(
+					LicenseCompareHelper.NORMALIZE_TOKENS.getOrDefault(token.toLowerCase(), token)) + 
+					"\\s*";
+		}
+	}
+	
+	static class RegexPattern extends RegexElement {
+		private String pattern;
+		
+		public RegexPattern(String pattern) {
+			this.pattern = pattern;
+		}
+		
+		public String getPattern() {
+			return pattern;
+		}
+		
+		@Override
+		public String toString() {
+			return "(" + pattern + ")"; // We always treat the pattern as a group
+		}
+
+		/**
+		 * @param pattern pattern to set
+		 */
+		public void setPattern(String pattern) {
+			this.pattern = pattern;
+		}
+	}
+
+	private String template;
```

Comment:
Rather than making this stateful (i.e. holding onto the template reference), I'd envisaged more of a stateless / functional design.  Basically receiving the String value of the template (or fragment) to be processed as an argument, and then returning a regex for that argument without retaining any state.

The advantage of that approach is that there will only ever need to be one instance of `TemplateRegexGenerator`, and it will have a low memory footprint (it won't retain references to templates, which are potentially large).

[edit] on further review, it seems like that would be incompatible with the `ILicenseTemplateOutputHandler` abstraction, so this comment might be moot.

Reply:
Retaining the state enabled me to remove a lot of nested code with local variables, so on balance I'm thinking the current design is simpler.